gossip/tx_scrambler_test.go

```diff
@@ -0,0 +1,597 @@
+					nonce:  1,
+					gas:    1,
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			res1 := test.entries
+			res2 := deepCopyEntries(res1)
+			// shuffle one array
+			shuffleEntries(res2)
+
+			res1 = getExecutionOrder(res1)
+			res2 = getExecutionOrder(res2)
+			if !reflect.DeepEqual(res1, res2) {
+				t.Error("slices have different order - algorithm is not deterministic")
+			}
+		})
+	}
+}
+
+func TestTxScrambler_GetExecutionOrder_SortIsDeterministic_RepeatedData(t *testing.T) {
+	tests := []struct {
+		name    string
+		entries []*scramblerEntry
+	}{
+		{
+			name: "repeated hashes",
+			entries: []*scramblerEntry{
+				{
+					hash:   common.Hash{1},
+					sender: common.Address{1},
+					nonce:  1,
+					gas:    1,
+				},
+				{
+					hash:   common.Hash{2},
+					sender: common.Address{2},
+					nonce:  2,
+					gas:    2,
+				},
+				{
+					hash:   common.Hash{3},
+					sender: common.Address{3},
+					nonce:  3,
+					gas:    3,
+				},
+				{
+					hash:   common.Hash{2},
+					sender: common.Address{2},
+					nonce:  2,
+					gas:    2,
+				},
+				{
+					hash:   common.Hash{1},
+					sender: common.Address{1},
+					nonce:  1,
+					gas:    1,
+				},
+			},
+		},
+		{
+			name: "repeated addresses",
+			entries: []*scramblerEntry{
+				{
+					hash:   common.Hash{1},
+					sender: common.Address{1},
+					nonce:  1,
+				},
+				{
+					hash:   common.Hash{2},
+					sender: common.Address{2},
+					nonce:  2,
+				},
+				{
+					hash:   common.Hash{3},
+					sender: common.Address{3},
+					nonce:  3,
+				},
+				{
+					hash:   common.Hash{4},
+					sender: common.Address{2},
+					nonce:  4,
+				},
+				{
+					hash:   common.Hash{5},
+					sender: common.Address{1},
+					nonce:  5,
+				},
+			},
+		},
+		{
+			name: "repeated addresses and nonces",
+			entries: []*scramblerEntry{
+				{
+					hash:   common.Hash{1},
+					sender: common.Address{1},
+					nonce:  1,
+					gas:    1,
+				},
+				{
+					hash:   common.Hash{2},
+					sender: common.Address{2},
+					nonce:  2,
+					gas:    2,
+				},
+				{
+					hash:   common.Hash{3},
+					sender: common.Address{3},
+					nonce:  3,
+					gas:    3,
+				},
+				{
+					hash:   common.Hash{4},
+					sender: common.Address{2},
+					nonce:  2,
+					gas:    4,
+				},
+				{
+					hash:   common.Hash{5},
+					sender: common.Address{1},
+					nonce:  1,
+					gas:    5,
+				},
+			},
+		},
+		{
+			name: "repeated addresses, nonces and gas",
+			entries: []*scramblerEntry{
+				{
+					hash:   common.Hash{1},
+					sender: common.Address{1},
+					nonce:  1,
+					gas:    1,
+				},
+				{
+					hash:   common.Hash{2},
+					sender: common.Address{2},
+					nonce:  2,
+					gas:    2,
+				},
+				{
+					hash:   common.Hash{3},
+					sender: common.Address{3},
+					nonce:  3,
+					gas:    3,
+				},
+				{
+					hash:   common.Hash{4},
+					sender: common.Address{2},
+					nonce:  2,
+					gas:    4,
```

Comment:
If I understand the test-case setup correctly, this should be a 2 and in the entry below a 1

Reply:
You are right, thanks